lang/java/avro/src/main/java/org/apache/avro/Schema.java

```diff
@@ -1867,47 +1781,70 @@ private static boolean isValidValue(Schema schema, JsonNode value) {
     }
   }
 
-  /**
-   * Parse named schema in order to fill names map. This method does not parse
-   * field of record/error schema.
-   *
-   * @param schema           : json schema representation.
-   * @param names            : map of named schema.
-   * @param currentNameSpace : current working name space.
-   * @return schema.
-   */
-  static Schema parseNamesDeclared(JsonNode schema, Names names, String currentNameSpace) {
+  /** @see #parse(String) */
+  static Schema parse(JsonNode schema, ParseContext context, String currentNameSpace) {
     if (schema == null) {
-      return null;
+      throw new SchemaParseException("Cannot parse <null> schema");
     }
-    if (schema.isObject()) {
-
-      String type = Schema.getOptionalText(schema, "type");
+    if (schema.isTextual()) { // name
+      return context.find(schema.textValue(), currentNameSpace);
+    } else if (schema.isObject()) {
+      Schema result;
+      String type = getRequiredText(schema, "type", "No type");
       Name name = null;
-
+      String space = null;
       String doc = null;
-      Schema result = null;
       final boolean isTypeError = "error".equals(type);
       final boolean isTypeRecord = "record".equals(type);
       final boolean isTypeEnum = "enum".equals(type);
       final boolean isTypeFixed = "fixed".equals(type);
-
       if (isTypeRecord || isTypeError || isTypeEnum || isTypeFixed) {
-        String space = getOptionalText(schema, "namespace");
+        space = getOptionalText(schema, "namespace");
         doc = getOptionalText(schema, "doc");
         if (space == null)
           space = currentNameSpace;
         name = new Name(getRequiredText(schema, "name", "No name in schema"), space);
       }
-      if (isTypeRecord || isTypeError) { // record
+      if (PRIMITIVES.containsKey(type)) { // primitive
+        result = create(PRIMITIVES.get(type));
+      } else if (isTypeRecord || isTypeError) { // record
+        List<Field> fields = new ArrayList<>();
         result = new RecordSchema(name, doc, isTypeError);
-        names.add(result);
+        context.put(result);
         JsonNode fieldsNode = schema.get("fields");
-
         if (fieldsNode == null || !fieldsNode.isArray())
           throw new SchemaParseException("Record has no fields: " + schema);
-        exploreFields(fieldsNode, names, name != null ? name.space : null);
-
+        for (JsonNode field : fieldsNode) {
```

Comment:
why not keep here a separate method for fields , and then, reduce complexity of this method ?

Reply:
Because the (semantic) complexity is in navigating the schema and fields twice, not in the technical method complexity. I'd be happy to split it in multiple methods to parse specific schema types though.

Reply:
I;'ve split the method to reduce complexity / increase readability.